### DISCUSSION TITLE
Cloud Storage Configuration for AWS in knowlg inQuiry Setup

### DISCUSSION THREAD

**Discussion by rajnishdargan** [Sunbird-Knowlg/Community#218] [Q&A]
Hello,

We have done the fresh setup of knowlg inQuiry setup for one of our project where we are using AWS.
Along with the knowlg inQuiry services, we are planning to use the knowlg and inQuiry editors and players.

We initially started with the [inQuiry Editor](https://github.com/Sunbird-inQuiry/editor) and successfully integrated it into our application. In the configuration for the Questionset Editor, there is a cloudStorage configuration that looks like this which is specific to azure:

```
const editorConfig = {
  context: {
   cloudStorageUrls: ['{{cloudStorageUrls path value}}'],
    cloudStorage: {
      provider: 'azure',
      presigned_headers: {
        'x-ms-blob-type': 'BlockBlob'
      }
    },
    // ...other configurations
  },
  config: {
    // ...other configurations
  }
}

```
The Questionset Editor internally uses the [@project-sunbird/sunbird-file-upload-library](https://www.npmjs.com/package/@project-sunbird/sunbird-file-upload-library)

I wanted to ask:

What should the cloudStorage configuration look like if we are using AWS?
How should the presigned_headers be structured for AWS in this case?
Any guidance or examples would be appreciated.

Thanks in advance!

@pallakartheekreddy @HarishGangula @rajeevsathish @vinukumar-vs @vaibhavbhuva
@coolbung @itsvick @AbhilashKD

**Comment by pallakartheekreddy** [comment]
@rajnishdargan Currently, sunbird-file-upload-library support uploading only Azure and OCI. To support AWS, you need to make code changes in below file.
https://github.com/Sunbird-Knowlg/sunbird-file-upload-library/blob/main/lib/index.js
Aftre that just you need change `cloudStorage` provider as `aws`

**Reply by rajnishdargan** [reply]
@pallakartheekreddy @HarishGangula 
Will changes be required in the [Content Editor](https://github.com/Sunbird-Knowlg/sunbird-content-editor) and [Generic Editor](https://github.com/Sunbird-Knowlg/sunbird-generic-editor) to support AWS? We are planning to use these editors as well, so I would like to know if any changes need to be made to the Generic and Content Editors to use them with AWS.

**Reply by pallakartheekreddy** [reply]
@rajnishdargan Once you're done above changes in `sunbird-file-upload-library` we need to publish new version and we need to use the same version in both the editors. To answer your question, yes we have to update library version in both the editors in `package.json` file.

**Comment by rajnishdargan** [comment]
Hi @pallakartheekreddy 

In the QuestionSet Editor, we updated the version of the [@project-sunbird/sunbird-file-upload-library](https://www.npmjs.com/package/@project-sunbird/sunbird-file-upload-library) from 1.0.2 to 1.0.4 to check if the latest code in the [sunbird-file-upload-library](https://github.com/Sunbird-Knowlg/sunbird-file-upload-library) works with AWS for file uploads.

In the updated code, I noticed the default handling case for CSP:
[DefaultUploader](https://github.com/Sunbird-Knowlg/sunbird-file-upload-library/blob/main/lib/index.js#L58)

We tested the AWS configuration for uploading images and videos. Surprisingly, the asset upload API worked successfully with the following config:

```
cloudStorage: {
   provider: 'aws',
   presigned_headers: {}
}
```

The pre-signed URL was generated,
The  artifactUrl and content_url looks like this:
```
artifactUrl: "{{cloud_storage_url}}/content/assets/do_2141582751952814081168/mountain.jpeg"
content_url: "{{cloud_storage_url}}/content/assets/do_2141582751952814081168/mountain.jpeg"
```

Could you please confirm if this setup is correct, or if there are any potential issues we might face later?

@HarishGangula @rajeevsathish @vinukumar-vs @vaibhavbhuva
@coolbung @itsvick @AbhilashKD
